data/tasks.js

```diff
@@ -0,0 +1,192 @@
+    return await this.getTaskById(newId.toString());
+  },
+
+  async updateTask(id, updatedTask) {
+    utils.checkObjectIdString(id);
+
+    const tasks = await tasksCollection();
+
+    const updatedTaskData = {};
+    updatedTask.title = updatedTask.title.trim();
+    updatedTask.dateAddedTo = updatedTask.dateAddedTo;
+    updatedTask.dateDueOn = updatedTask.dateDueOn;
+    updatedTask.textBody = updatedTask.textBody.trim();
+    updatedTask.tag = updatedTask.tag.trim().toLowerCase();
+    updatedTask.checked = updatedTask.checked;
+
+    if (updatedTask.title) {
+      utils.validateStringInput(updatedTask.title, "title");
+      updatedTaskData.title = updatedTask.title;
+    } else {
+      throw new Error("You must provide a title for the task.");
+    }
+
+    if (updatedTask.textBody) {
+      utils.validateStringInput(updatedTask.textBody, "textBody");
+      updatedTaskData.textBody = updatedTask.textBody;
+    } else {
+      throw new Error("You must provide a textBody for the task.");
+    }
+    if (updatedTask.dateAddedTo) {
+      utils.validateDate(updatedTask.dateAddedTo, "dateAddedTo");
+      updatedTaskData.dateAddedTo = new Date(updatedTask.dateAddedTo);
+    } else {
+      throw new Error("You must provide a dateAddedTo for the task.");
+    }
+
+    if (updatedTask.dateDueOn) {
+      utils.validateDate(updatedTask.dateDueOn, "dateDueOn");
+      updatedTaskData.dateDueOn = new Date(updatedTask.dateDueOn);
+    } else {
+      throw new Error("You must provide a dateDueOn for the task.");
+    }
+
+    if (updatedTask.priority) {
+      utils.validatePriority(updatedTask.priority);
+      updatedTaskData.priority = updatedTask.priority;
+    } else {
+      throw new Error("You must provide a priority for the task.");
+    }
+
+    if (updatedTask.tag) {
+      utils.validateStringInput(updatedTask.tag, "tag");
+      updatedTaskData.tag = updatedTask.tag;
+    } else {
+      throw new Error("You must provide a tag for the task.");
+    }
+
+    if (typeof updatedTask.checked !== "undefined") {
+      utils.validateBooleanInput(updatedTask.checked, "checked");
+      updatedTaskData.checked = updatedTask.checked;
+    } else {
+      throw new Error("You must provide a checked value for the task.");
+    }
+
+    const updateInfo = await tasks.updateOne(
+      { _id: new ObjectId(id) },
+      { $set: updatedTaskData }
+    );
+    if (updateInfo.modifiedCount === 0) {
+      throw new Error("Could not update task successfully");
+    }
+
+    return await this.getTaskById(id);
+  },
+
+  async removeTask(id) {
+    utils.checkObjectIdString(id);
+
+    const tasks = await tasksCollection();
+    const task = await this.getTaskById(id);
+    const deletionInfo = await tasks.deleteOne({ _id: ObjectId(id) });
+    if (deletionInfo.deletedCount === 0) {
+      throw new Error(`Could not delete task with ID ${id}`);
+    }
+
+    return task;
+  },
+
+  async getAllTasks(userId) {
+    utils.checkObjectIdString(userId);
```

Comment:
please trim the id

Reply:
checkObjectIdString(stringObjectId) {
    this.validateStringInput(stringObjectId, "objectID");
    stringObjectId = stringObjectId.trim();
    if (!ObjectId.isValid(stringObjectId)) {
      throw new Error("object id is not valid");
    }
    its doing it..